contents/javascript/Javascript-framework-Redux.md

```diff
@@ -0,0 +1,228 @@
+```
+
+A careful eye would notice that the above code block examples on reducers are not pure functions. If you're interested, you may see how the reducer should actually be written in the [official Redux site](https://redux.js.org/basics/reducers/#source-code). The code in the link includes some Redux API, and Javascript methods such as [`.map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) and the [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) to make the functions pure. But knowing the actual syntax for Redux is not the main focus of this introductory article.
+
+Finally, take note that there are no direct setters to the store, and the store is read-only. The only way the **store** is updated is through a strict unidirection data flow where: 
+1. An **Action** is dispatched and,
+1. Together with the current state read from the **Store**, is passed into a **Reducer**, which returns the new state.  
+
+## Why Use Redux? 
+
+<sup>[[1]](#footnote1)</sup> <sup>[[2]](#footnote2)</sup> 
+
+Now that we know what Redux is, let us take a look at how the structure of Redux and its concepts make possible some benefits.
+
+1. **State is made predictable**<br/>
+In Redux, the single immutable store is the single source of truth for application state. Neither the view of the application, nor network callbacks directly write to it. Changes to the state are via actions, which Redux centralizes and keep in strict order via the unidirection data flow. Also, reducers are pure functions, always producing the same result for the same state and action. 
+<br/><br/>Hence application state is updated in a predictable and structured manner. Having a predictable state, makes it easier to implement functionality that would be otherwise difficult, such as <tooltip content="The general idea is to store the history of the state in the state! To read more, click the link.">[Undo and Redo](https://redux.js.org/recipes/implementing-undo-history/)</tooltip>.
+
+2. **Easier to debug and inspect**<br/>
+Since actions and state are plain objects, it is possible to serialize them and log them. [Redux DevTools](https://github.com/reduxjs/redux-devtools) provides first class support for application debugging. You can view a log of actions that took place, making it easy to understand errors in code. You can also print the store <tooltip content="for example using `console.log(store.getState())`">within the code itself</tooltip>. The Redux DevTools, used in the browser, also provides features such as a tree visualization of the application state, as well a time travel ability to replay state changes.
+
+![Redux Devtool](https://user-images.githubusercontent.com/7957859/48663602-3aac4900-ea9b-11e8-921f-97059cbb599c.png)
+
+<center>
+
+_Redux DevTools_ <sup>[image source](https://user-images.githubusercontent.com/7957859/48663602-3aac4900-ea9b-11e8-921f-97059cbb599c.png)</sup>
```

Comment:
It's not common to give image source as a super script. Give as normal text?

Reply:
Seems there is a lot of white space below the image.

Reply:
That white space can't be helped, it's part of the original image.

Reply:
Is this your own image? In that case, no need to cite source. Also, can crop out the empty area.
If not your own image, credit the provider (e.g., original article) rather than the source URL.

Reply:
I see. The image is from the redux devtool's github readme. https://github.com/reduxjs/redux-devtools 

So should i just credit redux-devtools directly?

Reply:
Yes, that is a more meaningful acknowledgement of the source as the image URL itself does not have any mention of the original author.